lib/timers.js

```diff
@@ -62,6 +62,37 @@ function listOnTimeout() {
   var diff, first, threw;
   while (first = L.peek(list)) {
     diff = now - first._idleStart;
+    // This handles the case of a timer that was created within a timers
+    // callback with the same timeout value. For instance, when processing the
+    // timer that would call `bar` in such code:
+    //
+    //   setTimeout(function foo() { setTimeout(function bar() {}, 0) }, 0);
+    //
+    // or
+    //
+    //   setTimeout(function foo() { setTimeout(function bar() {}, 500) }, 500);
+    //
+    // We want to make sure that newly added timer fires in the next turn of the
+    // event loop at the earliest. So even if it's already expired now,
+    // reschedule it to fire later.
+    //
+    // At that point, it's not necessary to process any other timer in that
+    // list, because any remaining timer has been added within a callback of a
+    // timer that has already been processed, and thus needs to be processed at
+    // the earliest not in the current tick, but when the rescheduled timer will
+    // expire.
+    //
+    // See: https://github.com/joyent/node/issues/25607
+    if (now <= first._idleStart) {
+      var timeRemaining = msecs - (Timer.now() - first._idleStart);
+      if (timeRemaining < 0) {
+        timeRemaining = 0;
+      }
+      debug(msecs + ' list wait because timer was added from another timer');
+      list.start(timeRemaining, 0);
+      return;
+    }
+
     if (diff < msecs) {
```

Comment:
I still don't understand how this won't catch it.

`diff` is guaranteed to be zero or negative here if `now <= first._idleStart` is true. (This might need to be `<=`?)

Is the entire bug that `list.start(msecs - diff, 0);` starts adding if the value os negative?

If so, the logic should be rolled into a negative check in the statement below, I think.